src/Dapr.Messaging/PublishSubscribe/DaprPublishSubscribeClient.cs

```diff
@@ -0,0 +1,64 @@
+﻿// ------------------------------------------------------------------------
+// Copyright 2024 The Dapr Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+namespace Dapr.Messaging.PublishSubscribe;
+
+/// <summary>
+/// The base implementation of a Dapr pub/sub client.
+/// </summary>
+public abstract class DaprPublishSubscribeClient
+{
+    /// <summary>
+    /// Dynamically subscribes to a Publish/Subscribe component and topic.
+    /// </summary>
+    /// <param name="pubsubName">The name of the Publish/Subscribe component.</param>
+    /// <param name="topicName">The name of the topic to subscribe to.</param>
+    /// <param name="options">Configuration options.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing the various messages returned by the subscription.</returns>
+    public abstract IAsyncEnumerable<TopicMessage> SubscribeAsync(string pubsubName, string topicName, DaprSubscriptionOptions options, CancellationToken cancellationToken);
```

Comment:
While I'm all for an `IAsyncEnumerable<T>` based method, I wonder if some users may find it a bit intimidating.  I still think a `IDisposable SubscribeAsync(string pubsubName, string topicName, DaprSubscriptionOptions, DaprTopicHandler handler)` style method would be useful and still align with Rx concepts.

In this case, `DaprTopicHandler` would be defined as something like `Func<TopicMessage, Func<Task>, Task>`, an async delegate that accepts a `TopicMessage` and an async "ack" method.  It would probably be pretty easy to have an extension method that just layers on top of the `IAsyncEnumerable` method.

Reply:
I think the approach you started with makes more sense in light of the sequential acknowledgement requirement. I describe the new approach [here](https://github.com/dapr/dotnet-sdk/pull/1346#discussion_r1760465212).